core/lib/zksync_core/src/vm_runner/storage.rs

```diff
@@ -0,0 +1,456 @@
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt::Debug,
+    marker::PhantomData,
+    sync::Arc,
+    time::Duration,
+};
+
+use anyhow::Context as _;
+use async_trait::async_trait;
+use multivm::{interface::L1BatchEnv, vm_1_4_2::SystemEnv, zk_evm_latest::ethereum_types::H256};
+use once_cell::sync::OnceCell;
+use tokio::sync::{watch, RwLock};
+use vm_utils::storage::L1BatchParamsProvider;
+use zksync_dal::{Connection, ConnectionPool, Core, CoreDal};
+use zksync_state::{
+    AsyncCatchupTask, PgOrRocksdbStorage, ReadStorageFactory, RocksdbStorageBuilder,
+    StateKeeperColumnFamily, StateValue,
+};
+use zksync_storage::RocksDB;
+use zksync_types::{block::MiniblockExecutionData, L1BatchNumber, L2ChainId, StorageKey};
+
+/// Data needed to re-execute an L1 batch.
+#[derive(Debug, Clone)]
+pub struct BatchData {
+    /// Parameters for L1 batch this data belongs to.
+    pub l1_batch_env: L1BatchEnv,
+    /// Execution process parameters.
+    pub system_env: SystemEnv,
+    /// List of miniblocks and corresponding transactions that were executed within batch.
+    pub miniblocks: Vec<MiniblockExecutionData>,
+}
+
+#[derive(Debug, Clone)]
+struct StorageData {
+    batch_data: BatchData,
+    storage_diffs: HashMap<StorageKey, StateValue>,
+    factory_deps: HashMap<H256, Vec<u8>>,
+}
+
+/// Functionality to fetch data about processed/unprocessed batches for a particular VM runner
+/// instance.
+#[async_trait]
+pub trait VmRunnerStorageLoader: Debug + Send + Sync + Clone + 'static {
+    /// Unique name of the VM runner instance.
+    fn name() -> &'static str;
+
+    /// Returns number of the next unprocessed L1 batch. Can return `None` if there are no batches
+    /// to be processed.
+    ///
+    /// # Errors
+    ///
+    /// Propagates DB errors.
+    async fn first_unprocessed_batch(
```

Comment:
I think `first_unprocessed_batch` is not needed here, it should always be equal to `latest_processed_batch` + 1.
And I suggested adding the method `last_ready_to_be_loaded_batch` which should be capable of controlling max batch that should be loaded by `StorageSyncTask`